File: problems/queue/queue.py
import time
from multiprocessing.dummy import Pool
from multiprocessing import Pool as PPool
import humanfriendly

def blah(j):
    sum = 0
    for i in range(10000):
        sum += i

def avg(lst):
    return sum(lst) / len(lst)

if __name__ == "__main__":
    size_of_queue = 100
    num_items = 1000
    num_samples = 100

    time_taken = []
    for _ in range(num_samples):
        start = time.time()
        pool = Pool(processes = 10)
        pool.map(blah, range(num_items))
        pool.close()

        time_taken.append(time.time() - start)

    print("Done (threading)")
    print(humanfriendly.format_timespan(avg(time_taken)))

    time_taken = []
    for _ in range(num_samples):
        start = time.time()
        pool = PPool(processes = 10)
        pool.map(blah, range(num_items))
        pool.close()

        time_taken.append(time.time() - start)

    print("Done (processes)")
    print(humanfriendly.format_timespan(avg(time_taken)))

File: problems/queue/queue.go
package main

import "fmt"

func do_work(done *chan int, j int, size_of_queue int) {
    sum := 0

    for i := 0; i < 10000; i++ {
		sum += i
	}

	*done <- j
}


func main() {
	size_of_queue := 10
    num_work := 1000

	my_queue := make(chan int, size_of_queue)

	for i := 0; i < num_work; i++ {
		go do_work(&my_queue, i, size_of_queue)
	}

    for i := 0; i < num_work; i++ {
    	fmt.Println(<- my_queue)
	}
}
